app/src/org/commcare/android/database/user/models/ACasePreV6Model.java

```diff
@@ -0,0 +1,107 @@
+import org.javarosa.core.util.externalizable.ExtUtil;
+import org.javarosa.core.util.externalizable.ExtWrapList;
+import org.javarosa.core.util.externalizable.ExtWrapMapPoly;
+import org.javarosa.core.util.externalizable.ExtWrapNullable;
+import org.javarosa.core.util.externalizable.ExtWrapTagged;
+import org.javarosa.core.util.externalizable.Externalizable;
+import org.javarosa.core.util.externalizable.PrototypeFactory;
+
+/**
+ * A model extension which reads Resource models from the
+ * Version 6 db, and writes them back into the Version 7
+ * db.
+ * 
+ * 
+ * NOTE: This updater is _NOT ROBUST AGAINST METADATA
+ * CHANGES_. If the Resource model metadata changes, this
+ * needs to be modified to reflect the V6/7 metadata.  
+ * 
+ * @author ctsims
+ *
+ */
+public class ACasePreV6Model extends ACase {
+    
+    /**
+     Blank constructor for deserialization only!
+    */
+    public ACasePreV6Model() {
+        
+    }
+    
+    /* (non-Javadoc)
+     * @see org.commcare.resources.model.Resource#readExternal(java.io.DataInputStream, org.javarosa.core.util.externalizable.PrototypeFactory)
+     */
+    @Override
+    public void readExternal(DataInputStream in, PrototypeFactory pf) throws IOException, DeserializationException {
+        typeId = ExtUtil.readString(in);
+        id = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+        name = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+        closed = ExtUtil.readBool(in);
+        dateOpened = (Date)ExtUtil.read(in, new ExtWrapNullable(Date.class), pf);
+        recordId = ExtUtil.readInt(in);
+        indices = (Vector<CaseIndex>)ExtUtil.read(in, new ExtWrapList(CaseIndexUpdater.class));
+        data = (Hashtable)ExtUtil.read(in, new ExtWrapMapPoly(String.class, true), pf);
+    }
+    
+    /*
+     * (non-Javadoc)
+     * @see org.javarosa.core.util.externalizable.Externalizable#writeExternal(java.io.DataOutputStream)
+     */
+    public void writeExternal(DataOutputStream out) throws IOException {
+        ExtUtil.writeString(out, typeId);
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(id));
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(name));
+        ExtUtil.writeBool(out, closed);
+        ExtUtil.write(out, new ExtWrapNullable(dateOpened));
+        ExtUtil.writeNumeric(out, recordId);
+        ExtUtil.write(out, new ExtWrapList(indices));
+        ExtUtil.write(out, new ExtWrapMapPoly(data));
+    }
+    
+    public static class CaseIndexUpdater extends CaseIndex {
+        
+        /*
+         * serialization only!
+         */
+        public CaseIndexUpdater() {
+            
+        }
+
+        /* (non-Javadoc)
+         * @see org.javarosa.core.util.externalizable.Externalizable#readExternal(java.io.DataInputStream, org.javarosa.core.util.externalizable.PrototypeFactory)
+         */
+        public void readExternal(DataInputStream in, PrototypeFactory pf) throws IOException, DeserializationException {
```

Comment:
It's the read from v6 and the write for v7, so we need to add one of the re-implementations. The nice part about writing them down is that on V8(or12)'s implementation the 'current' object's becomes wrong, so this would have to get changed anyway, which is an invisible/dangerous requirement. This makes the object a safe "v6"/"v7" translation object robust against future changes.